Add SSL Support - completion of PR#5

ElvinEfendi opened the pull request:
This PR is based on the work done at https://github.com/openresty/lua-resty-upstream-healthcheck/pull/5.

It adds SSL support and addresses all the comments(tests and session reuse) made in the original PR.

One significant modification to the original PR is that now, the health check will fail if `ssl_verify = true` and the certificate verification fails.

/r @agentzh

sitano committed:
add ssl support

sitano committed:
reuse ctx.type to allow https

ElvinEfendi committed:
Merge branch 'master' into ssl_support

ElvinEfendi committed:
a blank line before the else line

ElvinEfendi committed:
session reuse

ElvinEfendi committed:
initial ssl testing

ElvinEfendi committed:
when ssl_verify, fail if certificate missmatch

ElvinEfendi committed:
added tests for SSL health check

ElvinEfendi committed:
removed trailing whitespace

ElvinEfendi committed:
assert that ssl session is reused

ElvinEfendi committed:
make ssl session reuse optional

ElvinEfendi committed:
document ssl_reuse_session flag

ElvinEfendi committed:
return test plan back with the adjusted number

ElvinEfendi committed:
Merge pull request #1 from Shopify/ssl_support

Ssl support

ElvinEfendi committed:
avoid permission requirement by using port > 1024

ElvinEfendi committed:
Merge pull request #2 from Shopify/use-different-port

avoid permission requirement by using port > 1024

ElvinEfendi commented:
ping

charlescng commented:
@agentzh Can you review?

agentzh commented on lib/resty/upstream/healthcheck.lua:
```diff
@@ -230,16 +230,26 @@ local function check_peer(ctx, id, peer, is_backup)
+    end
+
+    if ctx.type == "https" then
+        session, err = sock:sslhandshake(ctx.session, name, ctx.ssl_verify)
```
Is `session` a global Lua variable? We should declare it as a `local`.

agentzh replied:
Better use the `lua-releng` script of `nginx-devel-utils` to check your Lua code. I'm getting the following errors by running this tool against your branch:

```
$ lua-releng
lib/resty/upstream/healthcheck.lua: 0.03 (0.03)
Checking use of Lua global variables in file lib/resty/upstream/healthcheck.lua...
	op no.	line	instruction	args	; code
	53	[237]	SETGLOBAL	11 -14	; session
	54	[238]	GETGLOBAL	11 -14	; session
	73	[244]	GETGLOBAL	11 -14	; session
Checking line length exceeding 80...
```

So apparently the `session` is a misused Lua global variable.

agentzh commented on t/sanity.t:
```diff
@@ -1372,3 +1372,316 @@ healthcheck: peer \[0:0::1\]:12356 was checked to be ok
+}
+
+lua_shared_dict healthcheck 1m;
+init_worker_by_lua '
```
We should use `init_worker_by_lua_block` here for newly added tests.

agentzh commented on t/sanity.t:
```diff
@@ -1372,3 +1372,316 @@ healthcheck: peer \[0:0::1\]:12356 was checked to be ok
+--- config
+    location = /t {
+        access_log off;
+        content_by_lua '
```
We should use `content_by_lua_block` in newly added tests. Please fix other similar places.

agentzh commented:
Also, please rebase to the latest git master.

agentzh commented:
@ElvinEfendi Thank you for your patch! Please take care of my comments above. Thanks!

ElvinEfendi committed:
make sure session is declared as local variable

ElvinEfendi committed:
make sure line lengths do not exceed 80

ElvinEfendi committed:
use new *_by_lua_block directive

ElvinEfendi committed:
Merge pull request #3 from Shopify/comply-with-lua-releng-requirements

Comply with lua releng requirements

ElvinEfendi commented:
@agentzh thanks for reviewing, I addressed your comments.

> Also, please rebase to the latest git master.

I don't see any conflict with the upstream master(there were some initially but those have been sycned to this branch already). And all the commits from it are already included in this branch. Why do you think this branch is not in sync with the upstream master?

ElvinEfendi committed:
bump version

ElvinEfendi committed:
add status_table API with tests

ElvinEfendi committed:
add missing comma

ElvinEfendi commented:
@agentzh any more comment on this PR?

ElvinEfendi committed:
let users to pass custom server_name to tcpsock:sslhandshake function

ElvinEfendi committed:
Merge pull request #5 from Shopify/custom-server-name-option

let users to pass custom server_name to tcpsock:sslhandshake function

ElvinEfendi commented:
By default `peer.name`(an entry defined inside upstream block as `<host>:<port>`) is being passed to `tcpsock:sslhandshake` function as `server_name` which is used to validate the server name specified in the server certificate sent from the remote when `ssl_verify` is `true`. Refer to https://github.com/openresty/lua-nginx-module#tcpsocksslhandshake for more information. This is a constraint when we want to share a single server certificate to do SSL health check between multiple peers. Therefore I added TEST 20 and 21 to make `ssl_verify` option's behaviour more clear and added a new functionality that lets users to override `peer.name` with a custom name set during the initialization. TEST 21 covers the new functionality. These changes are added in commit: https://github.com/openresty/lua-resty-upstream-healthcheck/pull/36/commits/ce88ef0ffc8e2230882bbb7bd83247a35e9061c6

ElvinEfendi committed:
Merge pull request #4 from Shopify/status-table-api

add status_table API

solio commented:
why this pr has not bean merged？
这个ssl support什么时候能合到主干上？

agentzh commented:
@solio This is still in my review queue. Sorry for the delay on my side.

agentzh commented:
@ElvinEfendi Will you please rebase to the latest git master? I'm seeing merge conflicts while rebasing it myself:

```
$ git rebase master
First, rewinding head to replay your work on top of it...
Applying: add ssl support
Using index info to reconstruct a base tree...
M	README.markdown
M	lib/resty/upstream/healthcheck.lua
Falling back to patching base and 3-way merge...
Auto-merging lib/resty/upstream/healthcheck.lua
CONFLICT (content): Merge conflict in lib/resty/upstream/healthcheck.lua
Auto-merging README.markdown
error: Failed to merge in the changes.
Patch failed at 0001 add ssl support
The copy of the patch that failed is found in: .git/rebase-apply/patch
```

Thanks! And sorry for the delay on my side.

ElvinEfendi commented:
> Will you please rebase to the latest git master? I'm seeing merge conflicts while rebasing it myself:

Hmm I don't know why does it show conflict. There's no commit in the upstream after https://github.com/openresty/lua-resty-upstream-healthcheck/commit/69751de70d5ac6ab84947e8e074b7a14715fdb02 which is already in downstream as well https://github.com/Shopify/lua-resty-upstream-healthcheck/commit/69751de70d5ac6ab84947e8e074b7a14715fdb02.

ElvinEfendi commented:
I'd also like to highlight that https://github.com/openresty/lua-resty-upstream-healthcheck/pull/36/commits/6731f959d146264e8ba1b5adbd41968deb0ae92e is not related to this PR, it is just included here because the downstream is Shopify fork's _master_.

I can create another branch in Shopify's fork without that commit and make another PR if you don't want to merge that last unrelated commit.

agentzh commented:
@ElvinEfendi Cool, thanks!

berlincount commented:
hmmm, may I ping on this PR? :)

pavelnemirovsky commented:
@agentzh when do u plan to merge it? Thx

agentzh commented:
Sorry, we haven't had the time to look at this. Been busy with other things with higher priority.

agentzh commented:
@dndx @spacewander Please help review this PR once you have a chance. Thanks!

igbun commented:
Hi,
Is there any update regarding this PR?

Locoking commented:
Dears, any update on this feature?

TWHOI commented:
+1 year :-/
Would be nice to have this merged.

doujiang24 commented:
@rainingmaster do you have time to take a look at this PR?

leandromoreira commented:
@agentzh is there any plan to merge this ?

leandromoreira commented:
Hi  @ElvinEfendi did you publish a rock with these changes? If you didn't, I'm thinking to keep a `lua-resty-upstream-healthcheck-ssl` repo patching your changes to the main branch of this repo, what do you think?

leandromoreira commented:
I copied this PR and released it as a rock https://github.com/globocom/lua-resty-upstream-healthcheck

ElvinEfendi commented:
I'd forgotten about this and not planning on continuing. If this is still useful for some, please fork away and make another PR.
